File: backend/app/infrastructure/models/accommodation.py
from beanie import Document
from pydantic import Field
from typing import Optional, List
from datetime import datetime

class Hall(Document):
    tenant_id: str
    name: str
    capacity: int
    gender: Optional[str] = None
    is_active: bool = True

    blocks: List[str] = []

    class Settings:
        name = "halls"

class Room(Document):
    tenant_id: str
    hall_id: str
    room_number: str
    room_type: str
    capacity: int

    occupied: int = 0
    students: List[str] = []
    is_active: bool = True

    class Settings:
        name = "rooms"

class Accommodation(Document):
    tenant_id: str
    student_id: str

    housing_type: str = "school_hostel"  # school_hostel, outside_hostel, private_renting
    hall_id: Optional[str] = None
    room_id: Optional[str] = None

    outside_hostel_name: Optional[str] = None
    outside_hostel_address: Optional[str] = None
    outside_hostel_contact: Optional[str] = None

    private_address: Optional[str] = None
    private_city: Optional[str] = None
    private_contact: Optional[str] = None

    allocation_date: datetime = Field(default_factory=datetime.utcnow)
    check_in_date: Optional[datetime] = None
    check_out_date: Optional[datetime] = None

    is_active: bool = True

    class Settings:
        name = "accommodations"

class MaintenanceRequest(Document):
    tenant_id: str
    hall_id: str
    room_id: Optional[str] = None

    reported_by: str
    issue_description: str

    status: str = "pending"
    assigned_to: Optional[str] = None

    created_date: datetime = Field(default_factory=datetime.utcnow)
    resolved_date: Optional[datetime] = None

    class Settings:
        name = "maintenance_requests"

File: frontend/src/types/accommodation.ts
export interface Hall {
  id: string
  name: string
  capacity: number
}

export interface Room {
  id: string
  hall_id: string
  room_number: string
  room_type: string
  capacity: number
  occupied: number
  is_active: boolean
}

export interface AllocateRoomRequest {
  student_id: string
  hall_id: string
  room_id: string
}

export interface MaintenanceRequestCreate {
  hall_id: string
  room_id?: string
  issue_description: string
}

export interface MaintenanceRequestItem {
  id: string
  hall_id: string
  room_id?: string
  issue_description: string
  status: string
  assigned_to?: string
  created_date?: string
  resolved_date?: string
}

export interface OccupancyHallBreakdown {
  hall_id: string
  hall_name: string
  room_count: number
  capacity: number
  occupied: number
}

export interface OccupancySummaryResponse {
  total_halls: number
  total_rooms: number
  total_capacity: number
  total_occupied: number
  vacancy_rate: number
  halls: OccupancyHallBreakdown[]
}

export interface RoomOccupantResponse {
  student_id: string
  hall_id: string
  room_id: string
  check_in_date?: string
  check_out_date?: string
  is_active: boolean
}

export interface HousingSelectionRequest {
  housing_type: "school_hostel" | "outside_hostel" | "private_renting"
  hall_id?: string
  room_id?: string
  outside_hostel_name?: string
  outside_hostel_address?: string
  outside_hostel_contact?: string
  private_address?: string
  private_city?: string
  private_contact?: string
}

export interface StudentHousingStatusResponse {
  student_id: string
  school_fee_paid: boolean
  hostel_fee_paid: boolean
  housing_status: "unassigned" | "school_hostel" | "outside_hostel" | "private_renting"
  hall_id?: string
  hall_name?: string
  room_id?: string
  room_number?: string
  outside_hostel_name?: string
  outside_hostel_address?: string
  outside_hostel_contact?: string
  private_address?: string
  private_city?: string
  private_contact?: string
}

